File: DesktopTools/macOS/Functions/DeviceMaintenance/DeviceTimestampShow.m
//
//  DeviceTimestampShow.m
//  DesktopTool
//
//  Created by Makoto Morita on 2024/01/26.
//
#import "DeviceTimestamp.h"
#import "DeviceTimestampShow.h"
#import "ToolLogFile.h"

@interface DeviceTimestampShow () <DeviceTimestampDelegate>
    // 上位クラスの参照を保持
    @property (nonatomic) id                            delegate;
    @property (nonatomic) DeviceTimestamp              *deviceTimestamp;

@end

@implementation DeviceTimestampShow

    - (instancetype)initWithDelegate:(id)delegate {
        self = [super initWithDelegate:delegate];
        if (self) {
            [self enableClickButtonDoProcess:false];
            [self setDeviceTimestamp:[[DeviceTimestamp alloc] initWithDelegate:self]];
        }
        return self;
    }

    - (void)DeviceTimestamp:(DeviceTimestamp *)deviceTimestamp didUpdateState:(bool)available {
        if (available) {
            [self enableClickButtonDoProcess:true];
        }
    }

    - (void)invokeProcessOnSubQueue {
        // 現在時刻照会処理を実行
        [[self deviceTimestamp] inquiry];
    }

    - (void)DeviceTimestamp:(DeviceTimestamp *)deviceTimestamp didNotifyResponseQuery:(bool)success withErrorMessage:(NSString *)errorMessage {
        if (success == false) {
            [self terminateCommand:false withMessage:errorMessage];
            return;
        }
        // 現在時刻文字列をログ出力
        [[ToolLogFile defaultLogger] info:[deviceTimestamp currentTimestampLogString]];
        // 現在時刻文字列を画面表示
        [self appendStatusText:[deviceTimestamp currentTimestampString]];
        // 画面に制御を戻す
        [self terminateCommand:true withMessage:nil];
    }

#pragma mark - 終了処理

    - (void)terminateCommand:(bool)success withMessage:(NSString *)message {
        // 終了メッセージを画面表示／ログ出力
        if (success) {
            [self LogAndShowInfoMessage:message];
        } else {
            [self LogAndShowErrorMessage:message];
        }
        [self pauseProcess:success];
    }

@end

File: DesktopTools/dotNET/DesktopToolApp/DesktopToolApp/Functions/DeviceMaintenance/DeviceTimestampShow.cs
﻿using AppCommon;
using System.Threading.Tasks;

namespace DesktopTool
{
    internal class DeviceTimestampShow : ToolDoProcess
    {
        public DeviceTimestampShow(string menuItemName) : base(menuItemName) { }

        protected override void InvokeProcessOnSubThread()
        {
            Task task = Task.Run(RetrieveCurrentTimestamp);
        }

        private void RetrieveCurrentTimestamp()
        {
            // 現在時刻照会処理を実行
            new DeviceTimestamp().Inquiry(NotifyResponseQueryHandler);
        }

        private void NotifyResponseQueryHandler(DeviceTimestamp sender, bool success, string errorMessage)
        {
            if (success == false) {
                // 失敗時はログ出力
                TerminateCommand(false, errorMessage);
                return;
            }

            // 現在時刻文字列をログ出力
            AppLogUtil.OutputLogInfo(sender.CurrentTimestampLogString);

            // 現在時刻文字列を画面表示
            FunctionUtil.DisplayTextOnApp(sender.CurrentTimestampString, ViewModel.AppendStatusText);

            // 画面に制御を戻す
            TerminateCommand(true, string.Empty);
        }

        //
        // 終了処理
        //
        private void TerminateCommand(bool success, string message)
        {
            // 終了メッセージを画面表示／ログ出力
            if (message.Length > 0) {
                if (success) {
                    LogAndShowInfoMessage(message);
                } else {
                    LogAndShowErrorMessage(message);
                }
            }

            // 画面に制御を戻す
            PauseProcess(success);
        }
    }
}
